lib/OpenQA/ExpandPlaceholder.pm

```diff
@@ -0,0 +1,53 @@
+# Copyright (C) 2019 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+
+package OpenQA::ExpandPlaceholder;
+
+use strict;
+use warnings;
+
+# replace %NAME% with $settings{NAME}
+sub expand_placeholders {
+    my ($settings) = @_;
+
+    for my $value (values %$settings) {
+        next unless defined $value;
+
+        my %visited_top_level_placeholders;
+
+        eval { $value =~ s/%(\w+)%/_expand_placeholder($settings, $1, \%visited_top_level_placeholders)/eg; };
+        if ($@) {
+            return "Error: $@";
+        }
```

Comment:
I think we had a discussion about exceptions in the chat.
Using exceptions consequently and catching them at the upmost level possible would be preferable, but we decided that we make this improvement later.
Still, I think `_expand_placeholder` should die instead of returning a string, because handling this is more work here, so we need that eval.

Reply:
Last weekend i made a [new exception handling mechanism](https://mojolicious.org/perldoc/Mojo/Exception#check) for Mojolicious. That might come in handy for this stuff in the future.

Reply:
Yes, in the future :) we should not yet rely on this new version

Reply:
@kraih @okurz So do I need to change the code to return a string?

Reply:
Not sure myself now. @kraih @perlpunk should @Amrysliu change something here or keep as is?

Reply:
IMHO keep it as it is

Reply:
Errors are presented to the user, right? I'm always very sceptical of doing that, even in cases like this were there is not much risk of leaking information the user shouldn't have. Maybe just to be sure make the die statements throw arrayrefs and rethrow non-arrayref exceptions? Maybe that's the pattern we should encourage until we have real exception objects.
```perl
eval { ... };
if (my $err = $@) {
  return ref $err eq 'ARRAY' ? "Error: @$err" : die $err;
}

sub _expand_placeholder {
  ...
  die ["The key $key contains a circular reference, its value is $settings->{$key}.\n"];
  ...
}
```
Think this would be setting a good example.